File: data_structure/1/python/matrix_mul.py
def matrix_mul(a, b):
    """
    矩阵相乘
    """
    mat = []
    for _ in range(len(a)):
        mat.append([0] * len(b[0]))

    for i in range(len(a)):
        for j in range(len(b[0])):
            row = a[i]
            col = get_col(b, j)
            mat[i][j] = vector_dot(row, col)
    return mat

def get_col(b, j):
    col = [0] * len(b)
    for i, row in enumerate(b):
        col[i] = row[j]
    return col

def vector_dot(v1, v2):
    return sum(v1[i] * v2[i] for i in range(len(v1)))

if __name__ == "__main__":
    mat = matrix_mul(
        [
            [1,2,3],
            [3,4,5],
        ],
        [
            [1,2,4],
            [3,4,5],
            [55,6,10],
        ]
    )
    print(mat)

File: data_structure/1/js/matrix_mul.js
function matrixMul(a, b) {
    let mat = []

    for ( let i in a ) {
        let row = a[i] 
        mat.push([])
        for ( let j in b[0] ) {
            let num = vectorDot(row, getCol(b, j))
            mat[i].push(num)
        }
    }

    return mat
}

function vectorDot(v1, v2) {
    let sum = 0
    for ( let i in v1 )
        sum += v1[i] * v2[i]
    return sum
}

function getCol(b, j) {
    col = []
    for ( let i=0; i < b.length; i++ ) {
        col.push(b[i][j]);
    }
    return col;
}

mat = matrixMul(
    [
        [1,2,3],
        [3,4,5],
    ],
    [
        [1,2,4],
        [3,4,5],
        [55,6,10],
    ]
)
console.log(mat)